### PR TITLE
40 hydraulic power gas

### PR DESCRIPTION
**Pull request by FJanssen-TNO**
Refactored part of the names such that hydraulic power pipe constraints could be moved to head_loss_class allowing a reuse of the constraints for gas.

### CODE REVIEW

File: tests/test_hydraulic_power.py [hunk 1]
```diff
@@ -238,6 +238,164 @@ def test_hydraulic_power(self):
...
+                        v_inspect_line_ind.append(i)
+            ind_check = 0
+            for k in range(len(v_inspect)-1):
+                if v_inspect_line_ind[k] == v_inspect_line_ind[k+1]:
+                    np.testing.assert_allclose(pipe_hp[k] * pipe_mass[k+1] / pipe_mass[k], pipe_hp[k+1])
+                elif v_inspect_line_ind[k] < v_inspect_line_ind[k+1]:
+                    np.testing.assert_array_less((pipe_hp[k] + 1e-6)* pipe_mass[k + 1] / pipe_mass[k],
+                                               pipe_hp[k + 1])
+                    ind_check += 1
+                elif v_inspect_line_ind[k] > v_inspect_line_ind[k + 1]:
```
Comment (jimrojerTNO) [inline tests/test_hydraulic_power.py:391]: I would suggest we make the test such that the flow always increases over time avoiding the last elif statement, and instead giving an error if this is not the case

File: tests/test_hydraulic_power.py [hunk 2]
```diff
@@ -238,6 +238,164 @@ def test_hydraulic_power(self):
...
+        v_inspect = results[f"{pipe}.GasOut.Q"] / solution.parameters(0)[f"{pipe}.area"]
+        v_points = [i * v_max / solution.gas_network_settings["n_linearization_lines"] for i in
+         range(solution.gas_network_settings["n_linearization_lines"] + 1)]
+
+        v_volumetric_flow = (np.asarray(v_points) * np.pi * pipe_diameter ** 2 / 4.0)
+
+    def test_hydraulic_power_gas_multi_demand(self):
+        """
+
+        """
```
Comment (jimrojerTNO) [inline tests/test_hydraulic_power.py:316]: check for the hydraulic power conservation at the node

File: tests/test_hydraulic_power.py [hunk 3]
```diff
@@ -238,6 +238,164 @@ def test_hydraulic_power(self):
...
+        pipe_wall_roughness = solution.energy_system_options()["wall_roughness"]
+
+        temperature = 20.0
+        pipe_length = solution.parameters(0)[f"{pipe}.length"]
+
+        v_inspect = results[f"{pipe}.GasOut.Q"] / solution.parameters(0)[f"{pipe}.area"]
+        v_points = [i * v_max / solution.gas_network_settings["n_linearization_lines"] for i in
+         range(solution.gas_network_settings["n_linearization_lines"] + 1)]
+
+        v_volumetric_flow = (np.asarray(v_points) * np.pi * pipe_diameter ** 2 / 4.0)
```
Comment (jimrojerTNO) [inline tests/test_hydraulic_power.py:311]: to finish the test with an check on the absolute value of the hydraulic power loss over a line

File: tests/test_hydraulic_power.py [hunk 4]
```diff
@@ -238,6 +238,164 @@ def test_hydraulic_power(self):
             atol=10.0,
         )
 
+    def test_hydraulic_power_gas(self):
+        """
```
Comment (jimrojerTNO) [inline tests/test_hydraulic_power.py:242]: Docstring

File: tests/test_hydraulic_power.py [hunk 5]
```diff
@@ -238,6 +238,164 @@ def test_hydraulic_power(self):
...
+        pipe_length = solution.parameters(0)[f"{pipe}.length"]
+
+        v_inspect = results[f"{pipe}.GasOut.Q"] / solution.parameters(0)[f"{pipe}.area"]
+        v_points = [i * v_max / solution.gas_network_settings["n_linearization_lines"] for i in
+         range(solution.gas_network_settings["n_linearization_lines"] + 1)]
+
+        v_volumetric_flow = (np.asarray(v_points) * np.pi * pipe_diameter ** 2 / 4.0)
+
+    def test_hydraulic_power_gas_multi_demand(self):
+        """
```
Comment (jimrojerTNO) [inline tests/test_hydraulic_power.py:314]: docstring

File: src/mesido/gas_physics_mixin.py [hunk 6]
```diff
@@ -132,6 +132,10 @@ def __init__(self, *args, **kwargs):
 
         self._gas_pipe_topo_pipe_class_map = {}
 
+        self.__gas_pipe_disconnect_var = {}
+        self.__gas_pipe_disconnect_var_bounds = {}
```
Comment (jimrojerTNO) [inline src/mesido/gas_physics_mixin.py:136]: Only create the map, comment out the var and bounds

File: src/mesido/gas_physics_mixin.py [hunk 7]
```diff
@@ -234,11 +239,11 @@ def _get_min_bound(bound):
                 self.__gas_flow_direct_bounds[flow_dir_var] = (0.0, 1.0)
 
             # Still to be added in the future
-            # if parameters[f"{pipe_name}.disconnectable"]:
-            #     disconnected_var = f"{pipe_name}__is_disconnected"
-            #     self._gas_pipe_disconnect_map[pipe_name] = disconnected_var
-            #     self.__gas_pipe_disconnect_var[disconnected_var] = ca.MX.sym(disconnected_var)
-            #     self.__gas_pipe_disconnect_var_bounds[disconnected_var] = (0.0, 1.0)
+            if parameters[f"{pipe_name}.disconnectable"]:
```
Comment (jimrojerTNO) [inline src/mesido/gas_physics_mixin.py:242]: comment this part again as we do not use the variables.

File: src/mesido/head_loss_class.py [hunk 8]
```diff
@@ -1178,3 +1185,137 @@ def _demand_head_loss_path_constraints(self, optimization_problem, _ensemble_mem
             )
 
         return constraints
+
+    def _pipe_hydraulic_power_path_constraints(
```
Comment (jimrojerTNO) [inline src/mesido/head_loss_class.py:1189]: let's do the same for head_loss

File: tests/test_hydraulic_power.py (hunk 1 above)
Comment (FJanssen-TNO) [reply tests/test_hydraulic_power.py:391]: done

File: tests/test_hydraulic_power.py (hunk 2 above)
Comment (FJanssen-TNO) [reply tests/test_hydraulic_power.py:316]: added

File: tests/test_hydraulic_power.py (hunk 4 above)
Comment (FJanssen-TNO) [reply tests/test_hydraulic_power.py:242]: done

File: tests/test_hydraulic_power.py (hunk 5 above)
Comment (FJanssen-TNO) [reply tests/test_hydraulic_power.py:314]: done

File: src/mesido/gas_physics_mixin.py (hunk 6 above)
Comment (FJanssen-TNO) [reply src/mesido/gas_physics_mixin.py:136]: done

File: src/mesido/gas_physics_mixin.py (hunk 7 above)
Comment (FJanssen-TNO) [reply src/mesido/gas_physics_mixin.py:242]: done

File: src/mesido/head_loss_class.py (hunk 8 above)
Comment (FJanssen-TNO) [reply src/mesido/head_loss_class.py:1436]: done

File: tests/test_hydraulic_power.py [hunk 9]
```diff
@@ -238,6 +240,243 @@ def test_hydraulic_power(self):
...
+
+        # balance of hydraulic power
+        pipes_demand = ["Pipe_7c53", "Pipe_c50f"]
+        pipes_source = ["Pipe_0e39", "Pipe_f1a4"]
+        balance = np.zeros(3)
+        for p in pipes_demand:
+            balance += results[f"{p}.GasIn.Hydraulic_power"]
+        for p in pipes_source:
+            balance -= results[f"{p}.GasOut.Hydraulic_power"]
+        np.testing.assert_allclose(balance, 0.0, atol=1e-6)
```
Comment (jimrojerTNO) [inline tests/test_hydraulic_power.py:473]: although it is/should be equivalent, I would like it if the variables on the gas_node port are checked as well

**Review by jimrojerTNO — APPROVED** [review]

**Pull request merged**